hybris/egl/ws.c

```diff
@@ -27,117 +27,135 @@
 pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
 
 static struct ws_module *ws = NULL;
+static char ws_name[32] = { 0 };
 
-static void _init_ws()
+static EGLBoolean ensureCorrectWs(const char * egl_platform)
 {
-	if (ws == NULL)
-	{
-		pthread_mutex_lock(&mutex);
-		if (ws != NULL) {
-			pthread_mutex_unlock(&mutex);
-			return;
-		}
-
-		char ws_name[2048];
-		char *egl_platform;
-
-		// Mesa uses EGL_PLATFORM for its own purposes.
-		// Add HYBRIS_EGLPLATFORM to avoid the conflicts
-		egl_platform=getenv("HYBRIS_EGLPLATFORM");
-
-		if (egl_platform == NULL)
-			egl_platform=getenv("EGL_PLATFORM");
-
-		if (egl_platform == NULL)
-			egl_platform = DEFAULT_EGL_PLATFORM;
-
-		// The env variables may be defined yet empty
-		if (!strcmp(egl_platform, ""))
-			egl_platform = DEFAULT_EGL_PLATFORM;
-
-		const char *eglplatform_dir = PKGLIBDIR;
-		const char *user_eglplatform_dir = getauxval(AT_SECURE)
-		                                   ? NULL
-		                                   : getenv("HYBRIS_EGLPLATFORM_DIR");
-		if (user_eglplatform_dir)
-			eglplatform_dir = user_eglplatform_dir;
-
-		snprintf(ws_name, 2048, "%s/eglplatform_%s.so", eglplatform_dir, egl_platform);
-
-		void *wsmod = (void *) dlopen(ws_name, RTLD_LAZY);
-		if (wsmod==NULL)
-		{
-			fprintf(stderr, "ERROR: %s\n\t%s\n", ws_name, dlerror());
-			assert(0);
-		}
-		ws = dlsym(wsmod, "ws_module_info");
-		assert(ws != NULL);
-		ws->init_module(&hybris_egl_interface);
+	return strcmp(egl_platform, ws_name) == 0;
+}
+
+/*
+ * ws_init() will be called from __eglHybrisGetPlatformDisplayCommon() to
+ * load the selected platform. That means the loading of ws is postponed until
+ * a display is opened. Ws functions which doesn't depend on having a display
+ * opened must cope with not having a ws.
+ *
+ * The function will return EGL_FALSE if the opened ws is not the requested
+ * one. It doesn't make sense to have multiple ws'es opened; Android EGL (and
+ * by extension our implementation) allow only one display to be opened.
+ */
+EGLBoolean ws_init(const char * egl_platform)
+{
+	if (ws != NULL)
+		return ensureCorrectWs(egl_platform);
 
+	pthread_mutex_lock(&mutex);
+	if (ws != NULL) {
 		pthread_mutex_unlock(&mutex);
+		return ensureCorrectWs(egl_platform);;
 	}
+
+	char ws_lib_path[2048];
+
+	const char *eglplatform_dir = PKGLIBDIR;
+	const char *user_eglplatform_dir = getauxval(AT_SECURE)
+										? NULL
+										: getenv("HYBRIS_EGLPLATFORM_DIR");
```

Comment:
Style nitpick, the indentation of these lines doesn't look good now, use the original formatting or less indentation.

Reply:
Opps. It happens because I set my editor to render a tab character as 4 spaces. That's fixed now.